lib/rules/jest/no-if.js

```diff
@@ -25,63 +25,60 @@ module.exports = {
         'This is usually an indication that you',
         'are attempting to test too much at once',
         'or not testing what you intend to.',
-        'Consider breaking the conditional statement out',
-        'into a separate test to resolve this error.',
+        'Consider writing a separate test for',
+        'each fork in the conditional statement.',
       ].join(' '),
     },
   },
 
   create(context) {
-    let callExpressionDepth = 0;
+    const state = {
+      testDepth: 0,
+      nestedDepth: 0,
+    };
 
-    return {
-      CallExpression(node) {
-        if (ignore(node)) {
-          return;
-        }
-        callExpressionDepth++;
-      },
-      IfStatement(node) {
-        if (callExpressionDepth === 0) {
-          return;
-        }
+    const changeNestedDepth = (increase) => () =>
+      inTest(state) && (increase ? state.nestedDepth++ : state.nestedDepth--);
 
-        context.report({
-          messageId: 'noIf',
-          node,
-        });
-      },
-      ConditionalExpression(node) {
-        if (callExpressionDepth === 0) {
-          return;
-        }
+    const changeDepth = (increase) => (node) =>
+      !testCallExpression(node) &&
+      (increase ? state.testDepth++ : state.testDepth--);
 
-        context.report({
-          messageId: 'noConditional',
-          node,
-        });
-      },
-      'CallExpression:exit': (node) => {
-        if (ignore(node)) {
-          return;
-        }
-        callExpressionDepth--;
-      },
+    const ifStatement = (messageId) => (node) =>
+      !ignoreStatement(state) &&
+      context.report({
+        messageId,
+        node,
+      });
+
+    return {
+      CallExpression: changeDepth(true),
+      FunctionExpression: changeNestedDepth(true),
+      ArrowFunctionExpression: changeNestedDepth(true),
+      IfStatement: ifStatement('noIf'),
+      ConditionalExpression: ifStatement('noConditional'),
+      'FunctionExpression:exit': changeNestedDepth(),
+      'ArrowFunctionExpression:exit': changeNestedDepth(),
+      'CallExpression:exit': changeDepth(),
```

Comment:
Not sure if me or the pattern, but I had to read through this a lot to understand what the `changeNestedDepth`/ `changeDepth`/ with/ without argument actually meant.

Reply:
Personally, I'd probably just keep a stack of call expressions (as an array), every time you enter push `true` or `false` for if it's a test or not, every time you exit pop. Then, checking for an invalid conditional is just checking if the current stack entry is `true` or `false`.